// seajs-build/lib/utils/fs_ext.js

var fs = require('fs');
var path = require('path');
var rimraf = require('./rimraf');


/**
 * rm -rf dir.
 */
exports.rmdirRF = function(dir) {
    rimraf.sync(dir);
};

exports.rmSync = function(dir) {
    rimraf.sync(dir);
};


var isWindows = process.platform === 'win32',
    windowsDriveRegExp = /^[a-zA-Z]\:\/$/;

var exists = exports.exists = function(path) {
    if (isWindows && path.charAt(path.length - 1) === '/' &&
        path.charAt(path.length - 2) !== ':') {
        path = path.substring(0, path.length - 1);
    }

    try {
        fs.statSync(path);
        return true;
    } catch (e) {
        return false;
    }
};

function mkDir(dir) {
    if (!exists(dir) && (!isWindows || !windowsDriveRegExp.test(dir))) {
        fs.mkdirSync(dir, 511);
    }
}

function mkFullDir(dir) {
    var parts = dir.split('/'),
        currDir = '',
        first = true;

    parts.forEach(function (part) {
        //First part may be empty string if path starts with a slash.
        currDir += part + '/';
        first = false;

        if (part) {
            mkDir(currDir);
        }
    });
}

/**
 * mkdir -s dir.
 */
exports.mkdirS = function(dir) {

    mkFullDir(dir)
};


var existsSync = exports.existsSync = function(filepath) {
  if (fs.existsSync) {
    return fs.existsSync(filepath);
  } else {
    return path.existsSync(filepath);
  }
};
